File: godfather/roles/neutral/serial_killer.py
from godfather.roles.mixins import SingleAction
from godfather.game.types import Attack, Defense, Priority
from godfather.factions import SerialKillerNeutral

DESCRIPTION = 'You may stab someone every night.'


class SerialKiller(SingleAction):
    name = 'Serial Killer'
    description = DESCRIPTION

    def __init__(self):
        super().__init__()
        self.faction = SerialKillerNeutral()
        self.action = 'stab'
        self.action_gerund = 'stabbing'
        self.action_priority = Priority.SERIAL_KILLER
        self.action_text = 'stab a player'

    def defense(self):
        return Defense.BASIC

    async def run_action(self, actions, player, target):
        if target.role.defense() >= Defense.BASIC:
            return
        pl_record = actions.record[target.user.id]
        pl_record['nightkill']['result'] = True
        pl_record['nightkill']['type'] = Attack.BASIC
        pl_record['nightkill']['by'].append(player)

    async def tear_down(self, actions, player, target):
        record = actions.record[target.user.id]['nightkill']
        success = record['result'] and player in record['by']

        if not success:
            return await player.user.send('Your target was too strong to kill!')
        await target.user.send('You were stabbed by a Serial Killer. You have died!')

File: src/lib/mafia/roles/neutral/Serial_Killer.ts
import SerialKillerFaction from '@mafia/factions/neutral/SerialKiller';
import { Attack, Defence } from '@mafia/managers/NightActionsManager';
import type Player from '@mafia/structures/Player';
import { ActionRole } from '../../structures/ActionRole';
import type { NightAction } from '../../managers/NightAction';
import KillerAction from '@root/lib/mafia/actions/common/KillerAction';

export default class Serial_Killer extends ActionRole {
	public name = 'Serial Killer';
	public faction = new SerialKillerFaction();
	public actions: NightAction[] = [new KillerAction(this, Attack.Basic, 'stab')];

	public constructor(player: Player) {
		super(player);
		this.description = this.game.t('roles/neutral:serialKillerDescription');
		// this.actionText = this.game.t('roles/actions:serialKillerText');
		// this.actionGerund = this.game.t('roles/actions:serialKillerGerund');
		// this.actionParticiple = this.game.t('roles/neutral:serialKillerParticiple');
	}

	public get defence() {
		return Defence.Basic;
	}
}

Serial_Killer.aliases = ['SK'];
Serial_Killer.categories = [...Serial_Killer.categories, 'Random Neutral', 'Neutral Killing', 'Evil'];
